### PR TITLE
[Plugin] Add track progress property to vehicles

### PR DESCRIPTION
**Pull request by Basssiiie**
Hey all, this is my draft attempt at adding a property for track progress to vehicles for plugins. This would be very useful for my RideVehicleEditor plugin to adjust spacing between vehicles, and maybe others can find some usefulness in it as well!

- Track progress can be read and set for Car-instances.
- ~~If the chosen value is beyond the maximum amount of the current track piece, the vehicle will go to the relative position on the next track piece.~~
- If the chosen value is negative, the vehicle will go to the relative position on the previous track piece.

What I'm still planning to do:
- [x] ~~Clamp track progress at the end of tracks (for non-complete circuits).~~
- [x] ~~Sometimes if it jumps to the next or previous track, the position is off by 1.~~
- [x] ~~Test it thoroughly to make sure it doesn't crash or mess up the game somehow (so far it doesn't seem to).~~
- [x] Run clang-format.

Update 21 dec:
- After talking with Duncan, the proposed implementation has been changed to modify the vehicles 'remaining distance' instead.
- Duncan provided the related #13607 update flag which this implementation uses.
- The 'track progress' setter has been removed and replaced by a 'move' method which moves the car a relative distance in either forward or backwards along the track.

A screenshot of it working ingame:
afbeelding

Right now it works surprisingly well and I haven't really found any issues so far. 

I'd love to hear feedback though. This is my first pull request and I'm not even sure if this is the best approach for it. Please let me know if this is something that would fit into the source code and if there is anything I should change.

### CODE REVIEW

**Comment by James103** [comment]
Have you tried running clang-format on your changes?

**Comment by Basssiiie** [comment]
> Have you tried running clang-format on your changes?

No I haven't, I was looking into how to do that after the checks failed. I'll fix that tonight. :)

**Comment by duncanspumpkin** [comment]
This wont be able to set the correct sprite for vehicles going through reverser elements or rotate elements like log flumes, heartline.

**Comment by duncanspumpkin** [comment]
I think it will need to hook in early to the UpdateMotion functions. The vehicle placement code uses UpdateMotion to place the vehicles and this ensures that all of the values are correct. I'm not too sure how you would achieve that though.

**Comment by Basssiiie** [comment]
@duncanspumpkin Thank you for your reply. I'll give the remaining_distance a shot. What does it exactly hold? The distance to travel within this frame? Or the distance to the next track piece? Or something else?

**Comment by duncanspumpkin** [comment]
From memory the units are the same as track progress. It's a sort of left over amount after a vehicle update when things don't quite line up

File: src/openrct2/scripting/ScEntity.hpp
Comment (tupaschoal) [inline src/openrct2/scripting/ScEntity.hpp:533]: Do we need this first invalidate, as nothing has been performed yet?

Comment (IntelOrca) [reply src/openrct2/scripting/ScEntity.hpp:533]: yes, otherwise it will only invalidate the screen at its new position and not its old position.

**Comment by Basssiiie** [comment]
@duncanspumpkin I realise I checked in your code from #13607 here as well. Should I remove it or will it merge fine after your PR has been merged?

File: distribution/openrct2.d.ts
Comment (IntelOrca) [inline distribution/openrct2.d.ts:1087]: The formatting is a bit strange here.

Comment (Basssiiie) [reply distribution/openrct2.d.ts:1087]: The formatting was fixed with commit 1cfaf49 right? Or did I miss something?

**Comment by duncanspumpkin** [comment]
> @duncanspumpkin I realise I checked in your code from #13607 here as well. Should I remove it or should it merge fine after your PR has been merged?

Could you rebase now that my code has been merged.

File: distribution/openrct2.d.ts
Comment (IntelOrca) [reply distribution/openrct2.d.ts:1087]: Yes, my browser probably hadn't updated to show the diff with that commit.

**Comment by Basssiiie** [comment]
> Could you rebase now that my code has been merged.

Sure! I'll do that tonight or tomorrow when I have time. :)

File: src/openrct2/ride/Vehicle.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:829]: I'm pretty sure you don't need to do any of this  moveInfo changes. UpdateTrackMotion should take care of it.

File: src/openrct2/ride/Vehicle.cpp
Comment (duncanspumpkin) [inline src/openrct2/ride/Vehicle.cpp:7653]: Any reason you are doing this rather than just setting `VEHICLE_UPDATE_FLAG_COLLISION_DISABLED` as well as `VEHICLE_UPDATE_FLAG_SINGLE_CAR_POSITION`

File: src/openrct2/scripting/ScEntity.hpp
Comment (duncanspumpkin) [inline src/openrct2/scripting/ScEntity.hpp:617]: I'm pretty sure `MoveRelativeDistance` will now take care of the invalidating.

File: src/openrct2/ride/Vehicle.cpp
Comment (Basssiiie) [reply src/openrct2/ride/Vehicle.cpp:7653]: I can do that yes, that sounds like a better idea. Will it not interfere with any other code if I clear the `VEHICLE_UPDATE_FLAG_COLLISION_DISABLED`-flag again after `UpdateTrackMotion`? I am not sure when and where this flag is used elsewhere.

File: src/openrct2/scripting/ScEntity.hpp
Comment (Basssiiie) [reply src/openrct2/scripting/ScEntity.hpp:617]: Yes you're right, I'll remove it. :)

File: src/openrct2/ride/Vehicle.cpp
Comment (Basssiiie) [reply src/openrct2/ride/Vehicle.cpp:829]: Yes you're right, I'll remove it. :)

File: src/openrct2/ride/Vehicle.cpp
Comment (duncanspumpkin) [reply src/openrct2/ride/Vehicle.cpp:7643]: I know it's used for vehicle placement (which is not an issue). Have a look see if it's set for anything else

Comment (Basssiiie) [reply src/openrct2/ride/Vehicle.cpp:7643]: The only case I can find is the Giga Coaster cablelift, where it is enabled permanently. I don't think `UpdateMotionCollisionDetection` is called for the cablelift though, so it seems safe to clear it. :)

**Comment by duncanspumpkin** [comment]
This looks reasonable. Assuming it all works fine on your plugin I've no issues with it. @IntelOrca do you have anything to add?

**Review by IntelOrca — APPROVED** [review]

**Comment by Basssiiie** [comment]
> This looks reasonable. Assuming it all works fine on your plugin I've no issues with it. @IntelOrca do you have anything to add?

Yes it works perfect in my plugin. :) Only quirk I could find is that modifying a car on a slope sometimes moves other vehicles on the train as well. (Like I mentioned on Discord.) 

Other than that I tested a lot of special rides like Reverser Coaster, Chairlift, Giga Coaster, Lift, Go Karts, Heartline Twister, Log Flume reversers etc. as well and they all work perfectly!

**Comment by IntelOrca** [comment]
Just wondering whether `move` is the best name for the method in both the source code and the plugin API. Maybe it should be `moveBy` or `travel` or `travelBy`.

**Comment by Basssiiie** [comment]
> Just wondering whether `move` is the best name for the method in both the source code and the plugin API. Maybe it should be `moveBy` or `travel` or `travelBy`.

I can change it to `moveBy` if that is preferred. That'd also make it more in line with `[Viewport.moveTo()](https://github.com/OpenRCT2/OpenRCT2/blob/1d536e2aadb7141704ae4eab6a80bc9a906e35f8/distribution/openrct2.d.ts#L1975)`. 

`travel` and `travelBy` sounds a bit more unconventional, but that's just my personal feeling. I'd not be against it if it's the final verdict.

Who wants to make the final choice? 😃

**Comment by Broxzier** [comment]
Just `move` doesn't really explain what the function does, especially since it only takes one argument. `travelBy` would make it more clear that it has to do with the track, though something even more verbose like `moveAlongTrack` would be even better IMO.

**Comment by IntelOrca** [comment]
Yeah, my thoughts are that move is likely to mean, move by a coordinate relative to the map rather than an amount relative to the track. So I think it should just have a different name in case we have another function called `moveBy` or `moveTo` which move it by an x, y, z axis.

**Comment by Basssiiie** [comment]
So `travelBy` is the final decision?

**Comment by IntelOrca** [comment]
@duncanspumpkin any personal suggestions?

**Comment by duncanspumpkin** [comment]
No opinions here

**Comment by Basssiiie** [comment]
The `move` method is now renamed to `travelBy`. :)

**Comment by IntelOrca** [comment]
Okay finally, you just need to increment the plugin API version number and add this to the changelog. Sorry, I should have mentioned that previously.

**Review by IntelOrca — APPROVED** [review]

**Comment by Basssiiie** [comment]
Great, no worries! I think it should be properly rebased now? And the plugin API version and changelog have been updated. :)

**Pull request merged**
